pkg/settings/provider.go

```diff
@@ -0,0 +1,147 @@
+package settings
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	harvesterv1 "github.com/harvester/harvester/pkg/apis/harvesterhci.io/v1beta1"
+	ctlharvesterv1 "github.com/harvester/harvester/pkg/generated/controllers/harvesterhci.io/v1beta1"
+)
+
+type Provider interface {
+	Get(name string) string
+	Set(name, value string) error
+	SetIfUnset(name, value string) error
+	SetAll(settings map[string]Setting) error
+}
+
+type SettingsProvider struct {
+	context       context.Context
+	settings      ctlharvesterv1.SettingClient
+	settingsCache ctlharvesterv1.SettingCache
+	fallback      map[string]string
+}
+
+func NewSettingsProvider(
+	ctx context.Context,
+	settings ctlharvesterv1.SettingClient,
+	settingsCache ctlharvesterv1.SettingCache,
+) SettingsProvider {
+	return SettingsProvider{
+		context:       ctx,
+		settings:      settings,
+		settingsCache: settingsCache,
+		fallback:      map[string]string{},
+	}
+}
+
+func (s *SettingsProvider) Get(name string) string {
+	value := os.Getenv(GetEnvKey(name))
+	if value != "" {
+		return value
+	}
+	obj, err := s.settingsCache.Get(name)
+	if err != nil {
+		val, err := s.settings.Get(name, v1.GetOptions{})
+		if err != nil {
+			return s.fallback[name]
+		}
+		obj = val
+	}
+	if obj.Value == "" {
+		return obj.Default
+	}
+	return obj.Value
+}
+
+func (s *SettingsProvider) Set(name, value string) error {
+	envValue := os.Getenv(GetEnvKey(name))
+	if envValue != "" {
+		return fmt.Errorf("setting %s can not be set because it is from environment variable", name)
+	}
+	obj, err := s.settings.Get(name, v1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	obj.Value = value
+	_, err = s.settings.Update(obj)
+	return err
+}
+
+func (s *SettingsProvider) SetIfUnset(name, value string) error {
+	obj, err := s.settings.Get(name, v1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	if obj.Value != "" {
+		return nil
+	}
+
+	obj.Value = value
+	_, err = s.settings.Update(obj)
+	return err
+}
+
+func (s *SettingsProvider) SetAll(settingsMap map[string]Setting) error {
+	fallback := map[string]string{}
+
+	for name, setting := range settingsMap {
+		key := GetEnvKey(name)
+		value := os.Getenv(key)
+
+		obj, err := s.settings.Get(setting.Name, v1.GetOptions{})
+		if errors.IsNotFound(err) {
+			newSetting := &harvesterv1.Setting{
+				ObjectMeta: v1.ObjectMeta{
+					Name: setting.Name,
+				},
+				Default: setting.Default,
+			}
+			if value != "" {
+				newSetting.Value = value
+			}
+			if newSetting.Value == "" {
+				fallback[newSetting.Name] = newSetting.Default
+			} else {
+				fallback[newSetting.Name] = newSetting.Value
+			}
+			_, err := s.settings.Create(newSetting)
+			if err != nil {
+				return err
+			}
```

Comment:
Now that we run settings controllers in the webhook server, will it be a race condition here?
Assume that :
A setting value is set in Harvester server env. Webhook server starts before Harvester server. It will create the setting with an empty value which is not expected.

Reply:
Hmm... That is a practical concern. If we choose not go this way, in order to resolve https://github.com/harvester/harvester/pull/1492#discussion_r742904953, we might need to duplicate the "use default value" logic for webhook.

Reply:
We already used a client to fetch the setting resource. Can we simply check the default field?

Reply:
Yep. I think we can close this PR and check the default here.
https://github.com/harvester/harvester/blob/979b7f81b941a1f8643994641c59f686b04bc426/pkg/webhook/resources/virtualmachine/mutator.go#L106-L122